load_test/xmpp_amqp.erl

```diff
@@ -0,0 +1,647 @@
+         {?CONNECTION_FAILURE(amqp), spiral},
+         {?CONNECTION_FAILURE(xmpp), spiral},
+         {?OPEN_CHANNEL_TIME, histogram},
+         {?OPEN_CHANNEL_FAILURE, spiral},
+         {?OPEN_CHANNELS, spiral}
+        ]).
+
+-behaviour(amoc_scenario).
+
+-export([start/1]).
+-export([init/0]).
+
+
+-type binjid() :: binary().
+-type role() :: amqp
+              | xmpp.
+
+%================================================
+% Global initalization
+%================================================
+
+required_vars() ->
+    [
+     xmpp_servers,
+     amoc_users,
+     rabbit_host,
+     rabbit_port,
+     rabbit_username,
+     rabbit_password
+    ].
+
+% @doc This is simple function, which checks, if all required variables
+% are available. Throws an arror otherwise.
+-spec check_vars() -> ok | no_return().
+check_vars() ->
+    NonDefinedVars = lists:foldl(fun (Var, Acc) ->
+                                         case amoc_config:get(Var) of
+                                             undefined -> [Var|Acc];
+                                             _ -> Acc
+                                         end
+                                 end,
+                                 [],
+                                 required_vars()),
+    case NonDefinedVars of
+        [] -> ok;
+        _ -> error({undefined_variables, NonDefinedVars})
+    end.
+
+-spec init() -> ok.
+init() ->
+    check_vars(),
+    init_metrics(),
+    ensure_amqp_connection_pool(),
+    % Setting up barrier, which stops all amoc users after they connect
+    % to either MongooseIM or RabbitMQ.
+    barrier_init(after_connection, amoc_config:get(amoc_users)),
+    ok.
+
+-spec start(amoc_scenario:user_id()) -> no_return().
+start(MyID) ->
+    case get_my_role(MyID) of
+        xmpp -> xmpp_start(MyID);
+        amqp -> amqp_start(MyID)
+    end.
+
+%================================================
+% Metrics functions
+%================================================
+
+init_metrics() ->
+    [begin
+         exometer:ensure(Name, Type, []),
+         exometer_report:subscribe(exometer_report_graphite,  Name, datapoints(Type), 10000)
+     end
+     ||
+     {Name, Type} <- ?METRICS].
+
+-spec datapoints(exometer_report:metric()) -> exometer_report:datapoints().
+datapoints(spiral) -> [one, count];
+datapoints(histogram) -> [mean, min, max, median, 95, 99, 999].
+
+%================================================
+% Assigning fuctions
+%================================================
+
+-define(IS_AMQP(ID), ID rem 10 == 0).
+-define(IS_XMPP(ID), ID rem 10 =/= 0).
+
+%% @doc This function returns role for specific user.
+-spec get_my_role(amoc_scenario:user_id()) -> role().
+get_my_role(ID) when ?IS_AMQP(ID) -> amqp;
+get_my_role(ID) when ?IS_XMPP(ID) -> xmpp.
+
+%% @doc This function returns a list of neighbours which connects via xmpp
+-spec xmpp_neighbors(amoc_scenario:user_id()) -> [amoc_scenario:user_id()].
+xmpp_neighbors(MyID) ->
```

Comment:
Why such complicated _algorithm_ when you can go as simple as this (made appropriately readable):

```erlang
lists:filter(Pred,
   lists:seq(erlang:max(0,Id-Prev), Id-1) ++ lists:seq(Id+1,erlang:min(Id+Prev,Limit))
).
```

Reply:
What does actually lazy mean here?

Reply:
I like `get_sequence` function, but I came with simpler solution:
```
-spec xmpp_neighbors(amoc_scenario:user_id()) -> [amoc_scenario:user_id()].
xmpp_neighbors(MyID) ->
    AmocUsers = amoc_config:get(amoc_users),

    PrevNeighbors = [ ID || ID <- lists:seq(1, MyID-1), ?IS_XMPP(ID) ],
    LimitedPrevNeighbors = lists:sublist(lists:reverse(PrevNeighbors), ?NUMBER_OF_PREV_NEIGHBOURS),

    NextNeighbors = [ ID || ID <- lists:seq(MyID+1, AmocUsers), ?IS_XMPP(ID) ],
    LimitedNextNeighbors = lists:sublist(NextNeighbors, ?NUMBER_OF_PREV_NEIGHBOURS),

    LimitedPrevNeighbors ++ LimitedNextNeighbors.
```
It works in the same way:
 - do not exceed boundaries - `1 .. amoc_config:get(amoc_users]`
 - filters out AMQP clients
 - returns ?NUMBER_OF_*_NEIGHBOURS ids, but does not exceed boundaries

```
amoc_users = 20.
-define(IS_AMQP(ID), ID rem 10 == 0).
-define(IS_XMPP(ID), ID rem 10 =/= 0).

1> xmpp_amqp:xmpp_neighbors(1).
[2,3,4,5]
2> xmpp_amqp:xmpp_neighbors(5).
[4,3,2,1,6,7,8,9]
3> xmpp_amqp:xmpp_neighbors(9).
[8,7,6,5,11,12,13,14]
4> xmpp_amqp:xmpp_neighbors(11).
[9,8,7,6,12,13,14,15]
5> xmpp_amqp:xmpp_neighbors(20).
[19,18,17,16]
6> xmpp_amqp:xmpp_neighbors(0).
** exception error: no function clause matching lists:seq(1,-1) (lists.erl, line 243)
     in function  xmpp_amqp:xmpp_neighbors/1 (/Users/bartoszszafran/projects/amoc/_build/default/lib/amoc/scenarios/xmpp_amqp.erl, line 195)
7> xmpp_amqp:xmpp_neighbors(21).
** exception error: no function clause matching lists:seq(22,20) (lists.erl, line 243)
     in function  xmpp_amqp:xmpp_neighbors/1 (/Users/bartoszszafran/projects/amoc/_build/default/lib/amoc/scenarios/xmpp_amqp.erl, line
```

What do you think @mentels?